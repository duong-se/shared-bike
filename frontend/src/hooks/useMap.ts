import { useCallback, useState } from 'react'

export type UseMapReturn = {
  mapRef: (node: HTMLDivElement | null) => void
  map?: google.maps.Map
}

export const useMap = (position: google.maps.LatLng): UseMapReturn => {
  const [map, setMap] = useState<google.maps.Map>()
  const mapRef = useCallback((node: HTMLDivElement | null) => {
    if (node !== null) {
      const map = new window.google.maps.Map(node, {
        center: position,
        zoom: 8,
        fullscreenControl: false,
      })
      setMap(map)
    }
  }, [])
  return {
    map,
    mapRef,
  }
}
