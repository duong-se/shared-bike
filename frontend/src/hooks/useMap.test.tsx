import { renderHook, waitFor, act } from '@testing-library/react'
import { initialize } from '@googlemaps/jest-mocks'
import { useMap } from './useMap'

describe('useMap', () => {
  beforeEach(() => {
    initialize()
  })
  it('should run and return map', async () => {
    const centerPosition = new google.maps.LatLng(50.119504, 8.638137)
    const { result } = renderHook(() => useMap(centerPosition))
    await waitFor(() => result.current.map)
    expect(result.current.map).toEqual(undefined)
    expect(result.current.mapRef).toEqual(expect.any(Function))
    act(() => {
      result.current.mapRef(document.createElement('div'))
    })
    await waitFor(() => {
      expect(result.current.map).toBeInstanceOf(google.maps.Map)
    })
  })

  it('should run and not return map', async () => {
    const centerPosition = new google.maps.LatLng(50.119504, 8.638137)
    const { result } = renderHook(() => useMap(centerPosition))
    await waitFor(() => result.current.map)
    expect(result.current.map).toEqual(undefined)
    expect(result.current.mapRef).toEqual(expect.any(Function))
    act(() => {
      result.current.mapRef(null)
    })
    await waitFor(() => {
      expect(result.current.map).toMatchSnapshot()
    })
  })
})
