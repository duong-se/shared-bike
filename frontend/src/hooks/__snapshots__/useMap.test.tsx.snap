// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useMap should run and not return map 1`] = `undefined`;
